### PR TITLE
Add game Grensholm

### LINKED ISSUE #414: Add Grensholm
image

Rules:

* Soldiers move like knights in chess, except you must move towards the opponent's side
* Once a soldier reach the opponent side they are promoted to a "king", which still moves like a knight but must now move back home
* Winner is first to get a king back home
* Both soldiers and kings capture by replacement

Links: 

* http://www.di.fc.ul.pt/~jpn/gv/grensholm.htm
* https://www.zillions-of-games.com/cgi-bin/zilligames/submissions.cgi/49215?do=show;id=1009
* http://blog.krawaller.se/algol4/#/game/grensholm/info

Implementation notes:

* The rules of Grensholm are super similar to [Ugly Duck](https://chessicals.com/games/ugly_duck/), so check out the definition of that game for lots of help!
* The UI in Chessicals doesn't work well with "portrait" rectangles, so we should rotate the board 90 degrees (so that it is 6x4 instead of 4x6)
* To make "chess horse movement" available we must add an offset definition to the board. This will generate all "horse links" between squares initially, so we don't have to calculate them during play. See for example [Chameleon](https://chessicals.com/games/chameleon/) to see how this is done
* To make a piece able to move like a chess horse, you would normally just use `dirs: "knight"`. But here we just want to make SOME of those dirs available (those who move us towards the opponent base, or back if we're a king). If you look in the file `modules/common/connections/dirConstants.ts`, you'll see that `"knight"` is just a shortcut for a range of offset dirs, just like `"ortho"` is a shortcut for `[1, 3, 5, 7]`. So you need to hand-pick the correct offsets from that list, depending on if we're a soldier or king, and if we're player 1 or 2.

Suggested iteration:

* Flow of the game would be sth like: `selectunit -> selectmovetarget -> move`, with a single generator `findmovetargets` triggered by `selectunit`.
* Start by just making the soldiers move like regular chess horsers (`dirs: "knight"`), and capture by replacement (use `stomp` for movement effect).
* Next add promotion to king if we end our move in opponent base
* Next make it so that soldiers have to move towards opp base and kings must move homewards
* Next add win condition for having a king in the own base! Pro tip: to test this, make a temporary change to the setup where you place a king within reach of the home row, that way you don't have to play through an almost full game just to see if it works!

### PR DESCRIPTION
**Pull request by wnr**
Closes #414

### CODE REVIEW

File: modules/content/material/games/grensholm/links.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export const links = {
+  "World of Abstract Games entry": "http://www.di.fc.ul.pt/~jpn/gv/grensholm.htm",
+  "Zillions of Games entry": "https://www.zillions-of-games.com/cgi-bin/zilligames/submissions.cgi/49215?do=show;id=1009",
+  "Krawaller Blog entry": "http://blog.krawaller.se/algol4/#/game/grensholm/info"
```
Comment (krawaller) [inline modules/content/material/games/grensholm/links.ts:4]: This old implementation is just embarrassing, let's leave out that link! 🙈

File: modules/games/definitions/grensholm/generators.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  "d3f2r-1",
+  "d5f2r-1"
+];
+
+const knightRightToLeftDirs = [
+  "d1f2r-1",
+  "d5f2r1",
+  "d7f2r1",
+  "d7f2r-1"
+];
```
Comment (krawaller) [inline modules/games/definitions/grensholm/generators.ts:20]: So, there is a hidden layer of abstraction hidden here - I have a `retemplate` command which I sometimes need to run when I make language changes, and that rewrites all game definition files using the value they exported. But, there I assume that there are no dependencies except the generated imports. 

In other words, we can't have local variables in the game definition files. Sorry! 😊 

I should really rework that, or at least add some kind of mechanism for variables or repeated expressions in general. But for now we just have to be non-DRY and inline the values. :/

**Review by krawaller — COMMENTED** [review]
So solid!! 🙌 

Just two minor things;

File: modules/content/material/games/grensholm/links.ts (hunk 1 above)
Comment (wnr) [reply modules/content/material/games/grensholm/links.ts:4]: Got it :D

File: modules/games/definitions/grensholm/generators.ts (hunk 2 above)
Comment (wnr) [reply modules/games/definitions/grensholm/generators.ts]: I see, I'll fix it :)

**Comment by krawaller** [comment]
Deployed! Thank you @wnr! ❤️

**Pull request merged**